### PR TITLE
Add offline support for version maps of `istioctl x upgrade`

### LINKED ISSUE #17912: Upgrade version maps should be available offline
Version maps specify which versions are supported from which operator. The source of truth is master but we need a local copy built into operator for offline upgrade cases.

### PR DESCRIPTION
**Pull request by elfinhe**
Fall back to using the built-in version map if the remote one is not available. + some minor fixes on output.

Version maps specify which versions are supported from which operator. The source of truth is master but we need a local copy built into operator for offline upgrade cases.
Resolve: https://github.com/istio/istio/issues/17912

### CODE REVIEW

File: cmd/mesh/manifest-versions.go [hunk 1]
```diff
@@ -105,8 +104,26 @@ func getVersionCompatibleMap(versionsURI string, binVersion *goversion.Version,
 		}
 	}
 	if myVersionMap == nil {
-		l.logAndFatal("This operator version ", binVersion.String(),
-			" was not found in the global manifestVersions map.")
+		return nil, fmt.Errorf("this operator version %v was not found in the global manifestVersions map", binVersion.String())
```
Comment (ostromart) [inline cmd/mesh/manifest-versions.go]: nit: use %s for string. %v is ok but not idiomatic.

File: cmd/mesh/manifest-versions.go [hunk 2]
```diff
@@ -105,8 +104,26 @@ func getVersionCompatibleMap(versionsURI string, binVersion *goversion.Version,
...
+
+func loadCompatibleMapFile(versionsURI string, l *logger) (b []byte, err error) {
+	if _, err = url.ParseRequestURI(versionsURI); err == nil {
+		b, err = httprequest.Get(versionsURI)
+		if err == nil {
+			return
+		}
+	}
+	l.logAndPrintf("Warning. Failed to retrieve the version map from the remote url: %v, error: %v."+
+		" Now fall back to the local file system.", versionsURI, err)
```
Comment (ostromart) [inline cmd/mesh/manifest-versions.go]: nit: "Warning: failed to retrieve the version map from the remote URL (%s ): %s. Falling back to internal version map."
Here and below.

File: cmd/mesh/manifest-versions.go [hunk 3]
```diff
@@ -105,8 +104,26 @@ func getVersionCompatibleMap(versionsURI string, binVersion *goversion.Version,
...
 	}
 	if myVersionMap == nil {
-		l.logAndFatal("This operator version ", binVersion.String(),
-			" was not found in the global manifestVersions map.")
+		return nil, fmt.Errorf("this operator version %v was not found in the global manifestVersions map", binVersion.String())
+	}
+	return myVersionMap, nil
+}
+
+func loadCompatibleMapFile(versionsURI string, l *logger) (b []byte, err error) {
```
Comment (ostromart) [inline cmd/mesh/manifest-versions.go]: The logic here is not quite right. We should try versionsURI as a get if it has a http: prefix. Otherwise try it as a local file path. 
Depending on which way we try it, emit a warning about either remote URL or file.

File: cmd/mesh/manifest-versions_test.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+			args: args{
+				versionsURI: "__nonexistent-versions.yaml",
+				binVersion:  goVer132,
+				l:           l,
+			},
+			want:    vm132,
+			wantErr: nil,
+		},
+		{
+			name: "read from github",
```
Comment (ostromart) [inline cmd/mesh/manifest-versions_test.go]: I suggest skipping the read from github unless you feel like implementing a server in the test so that it's hermetic. We don't want to be calling out to a real URL in a test.
Instead, I suggest pointing it to a file in testdata. This will allow you also to test the negative case if path is incorrect, should get an error.

File: cmd/mesh/manifest-versions_test.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	goVer131000, _ := goversion.NewVersion("1.3.1000")
+	goVer132, _ := goversion.NewVersion("1.3.2")
+	scVer132, _ := goversion.NewConstraint(">=1.3.0,<=1.3.2")
+	rcVer132, _ := goversion.NewConstraint("1.3.2")
+	vm132 := &version.CompatibilityMapping{
+		OperatorVersion:          goVer132,
+		SupportedIstioVersions:   scVer132,
+		RecommendedIstioVersions: rcVer132,
+	}
```
Comment (ostromart) [inline cmd/mesh/manifest-versions_test.go]: this should probably come from the real versions map.
see version/version_test.go to see how it's read.

File: cmd/mesh/manifest-versions.go (hunk 1 above)
Comment (elfinhe) [reply cmd/mesh/manifest-versions.go]: Fixed.

File: cmd/mesh/manifest-versions.go (hunk 2 above)
Comment (elfinhe) [reply cmd/mesh/manifest-versions.go]: Sure, fixed.

File: cmd/mesh/manifest-versions.go (hunk 3 above)
Comment (elfinhe) [reply cmd/mesh/manifest-versions.go]: Yes, I fixed the logic here. Need a double-check. @ostromart

File: cmd/mesh/manifest-versions_test.go (hunk 4 above)
Comment (elfinhe) [reply cmd/mesh/manifest-versions_test.go]: Sure, will use test data instead.

File: cmd/mesh/manifest-versions_test.go (hunk 5 above)
Comment (elfinhe) [reply cmd/mesh/manifest-versions_test.go]: sure, fixed.

File: cmd/mesh/manifest-versions.go (hunk 3 above)
Comment (ostromart) [reply cmd/mesh/manifest-versions.go]: lgtm

File: cmd/mesh/manifest-versions.go [hunk 6]
```diff
@@ -105,8 +103,25 @@ func getVersionCompatibleMap(versionsURI string, binVersion *goversion.Version,
...
-		l.logAndFatal("This operator version ", binVersion.String(),
-			" was not found in the global manifestVersions map.")
+		return nil, fmt.Errorf("this operator version %s was not found in the version map", binVersion.String())
+	}
+	return myVersionMap, nil
+}
+
+func loadCompatibleMapFile(versionsURI string, l *logger) (b []byte, err error) {
+	if util.IsHTTPURL(versionsURI) {
+		b, err = httprequest.Get(versionsURI)
```
Comment (ostromart) [inline cmd/mesh/manifest-versions.go]: nit: it's best to avoid named return values where possible for improved readability through the smallest possible scoping, and also to remain consistent with project style. also, it's normal to redefine err as a local var where possible. 
here, i propose changing to:
if b, err :=  httprequest.Get(versionsURI); err == nil {
  return b, nil
}

Comment (elfinhe) [reply cmd/mesh/manifest-versions.go]: Sure, fixed.

**Review by ostromart — APPROVED** [review]
